### PR TITLE
TF-3826 DNS look up error handling

### PR DESCRIPTION
**Pull request by tddang-linagora**
## Issue
- #3826 

### No internet error message

[no-internet.webm](https://github.com/user-attachments/assets/aea840dc-87bd-466c-80ed-3e8008a52c86)

### DNS lookup failed, webfinger guessed successfully

https://github.com/user-attachments/assets/857a714c-64a3-486d-bbb5-15ba9b00dbe0

### DNS lookup failed, webfinger guessing failed

[guess-webfinger-fail.webm](https://github.com/user-attachments/assets/935d1f5d-e9b9-40e3-a5d7-ec645216d4dd)

### CODE REVIEW

File: lib/features/login/presentation/widgets/login_message_widget.dart [hunk 1]
```diff
@@ -47,6 +48,10 @@ class LoginMessageWidget extends StatelessWidget {
         child: Text(
           viewState.fold(
             (failure) {
+              if (failure is FeatureFailure && failure.exception is NoNetworkError) {
+                return AppLocalizations.of(context).youAreOffline;
+              }
+
```
Comment (tddang-linagora) [inline lib/features/login/presentation/widgets/login_message_widget.dart:54]: Specific message for `DNSLookupToGetJmapUrlFailure`

File: lib/features/login/domain/model/oidc_guessing_origin.dart [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  final String origin;
+
+  String url(String email) {
+    final emailDomain = email.split('@')[1];
+    return switch (this) {
+      OidcGuessingOrigin.empty => 'https://$emailDomain',
+      _ => 'https://$origin.$emailDomain',
+    };
+  }
```
Comment (hoangdat) [inline lib/features/login/domain/model/oidc_guessing_origin.dart]: write unit test for this

Comment (tddang-linagora) [reply lib/features/login/domain/model/oidc_guessing_origin.dart]: Tests are already written.

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      yield Right(TryingGuessingWebFinger());
+      final results = await Future.wait(oidcRequests.map(
+        (oidcRequest) async {
+          try {
+            return await _authenticationOIDCRepository.checkOIDCIsAvailable(
+              oidcRequest,
+            );
+          } catch (_) {
+            return null;
+          }
```
Comment (dab246) [inline lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]: Split function for it

File: lib/features/login/presentation/extensions/generate_oidc_guessing_urls.dart [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import 'package:get/get_utils/src/get_utils/get_utils.dart';
+import 'package:model/oidc/request/oidc_request.dart';
+import 'package:tmail_ui_user/features/login/domain/model/oidc_guessing_origin.dart';
+import 'package:tmail_ui_user/features/login/presentation/login_controller.dart';
+
+extension GenerateOidcGuessingUrls on LoginController {
+  List<OIDCRequest> generateOidcGuessingUrls(String email) {
+    if (!GetUtils.isEmail(email)) {
```
Comment (dab246) [inline lib/features/login/presentation/extensions/generate_oidc_guessing_urls.dart]: Please use `isEmailAddressValid` in EmailUtils

File: lib/features/login/domain/model/oidc_guessing_origin.dart [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  const OidcGuessingOrigin({required this.origin});
+
+  final String origin;
+
+  String url(String email) {
+    final emailDomain = email.split('@')[1];
+    return switch (this) {
+      OidcGuessingOrigin.empty => 'https://$emailDomain',
+      _ => 'https://$origin.$emailDomain',
```
Comment (dab246) [inline lib/features/login/domain/model/oidc_guessing_origin.dart]: What about If origin is only whitespace?

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart (hunk 3 above)
Comment (tddang-linagora) [reply lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]: Done

File: lib/features/login/presentation/extensions/generate_oidc_guessing_urls.dart (hunk 4 above)
Comment (tddang-linagora) [reply lib/features/login/presentation/extensions/generate_oidc_guessing_urls.dart]: Done

File: lib/features/login/domain/model/oidc_guessing_origin.dart (hunk 5 above)
Comment (tddang-linagora) [reply lib/features/login/domain/model/oidc_guessing_origin.dart]: Updated

File: lib/features/login/domain/model/oidc_guessing_origin.dart [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+  const OidcGuessingOrigin({required this.origin});
+
+  final String origin;
+
+  String url(String email) {
+    final emailDomain = email.split('@')[1];
+    return switch (this) {
+      OidcGuessingOrigin.empty => 'https://$emailDomain',
+      _ => origin.isEmpty
```
Comment (dab246) [inline lib/features/login/domain/model/oidc_guessing_origin.dart]:
```suggestion
      _ => origin.trim().isEmpty
```

Comment (tddang-linagora) [reply lib/features/login/domain/model/oidc_guessing_origin.dart]: Done

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      } else {
+        yield Right(TryGuessingWebFingerSuccess(availableOidcResponses.first));
+      }
+    } catch (e) {
+      logError('$runtimeType::execute(): Exception = $e');
+      yield Left(TryGuessingWebFingerFailure(exception: e));
+    }
+  }
+
+  Future<OIDCResponse?> _getWebFingerFromOidcRequest(
```
Comment (hoangdat) [inline lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]:
```suggestion
  Future<OIDCResponse?> _checkOIDCAvailableFromOidcRequest(
```

File: lib/features/login/presentation/widgets/login_message_widget.dart (hunk 1 above)
Comment (hoangdat) [reply lib/features/login/presentation/widgets/login_message_widget.dart:54]: maaybe you forget it @tddang-linagora

File: test/features/login/domain/model/oidc_guessing_origin_test.dart [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:tmail_ui_user/features/login/domain/model/oidc_guessing_origin.dart';
+
+void main() {
+  group('OidcGuessingOrigin enum', () {
```
Comment (hoangdat) [inline test/features/login/domain/model/oidc_guessing_origin_test.dart:5]: how about the case with email have double @, or more edge cases?

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart (hunk 7 above)
Comment (tddang-linagora) [reply lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]: Done

File: test/features/login/domain/model/oidc_guessing_origin_test.dart (hunk 8 above)
Comment (tddang-linagora) [reply test/features/login/domain/model/oidc_guessing_origin_test.dart:5]: Done

File: lib/features/login/presentation/widgets/login_message_widget.dart (hunk 1 above)
Comment (tddang-linagora) [reply lib/features/login/presentation/widgets/login_message_widget.dart:54]: Separate urgent exception handled. This is when `ConnectionError` is thrown.

[untitled.webm](https://github.com/user-attachments/assets/992fea7a-0e14-4909-89a3-dcf21ab3cf7d)

Comment (hoangdat) [reply lib/features/login/presentation/widgets/login_message_widget.dart:54]: It means, connection happen, trying guess mechanism be triggered?

File: test/features/login/domain/model/oidc_guessing_origin_test.dart (hunk 8 above)
Comment (hoangdat) [reply test/features/login/domain/model/oidc_guessing_origin_test.dart:5]: still not have this case with double @ in test

File: lib/features/login/presentation/widgets/login_message_widget.dart (hunk 1 above)
Comment (tddang-linagora) [reply lib/features/login/presentation/widgets/login_message_widget.dart:54]: correct

File: test/features/login/domain/model/oidc_guessing_origin_test.dart (hunk 8 above)
Comment (tddang-linagora) [reply test/features/login/domain/model/oidc_guessing_origin_test.dart:5]: Updated

**Review by dab246 — APPROVED** [review]
- [ ] Please rebase it

File: lib/features/login/domain/model/oidc_guessing_origin.dart
Comment (hoangdat) [inline lib/features/login/domain/model/oidc_guessing_origin.dart:1]: should have an ADR for the guessing mechanism. Thank you

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  final AuthenticationOIDCRepository _authenticationOIDCRepository;
+
+  Stream<Either<Failure, Success>> execute(
+    List<OIDCRequest> oidcRequests,
+  ) async* {
+    try {
+      yield Right(TryingGuessingWebFinger());
+      final results = await Future.wait(oidcRequests.map(
+        _checkOIDCAvailableFromOidcRequest,
+      ));
```
Comment (hoangdat) [inline lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]: why not stop the guessing whenever found jmap?

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import 'package:model/oidc/response/oidc_response.dart';
+import 'package:tmail_ui_user/features/login/domain/repository/authentication_oidc_repository.dart';
+import 'package:tmail_ui_user/features/login/domain/state/try_guessing_web_finger_state.dart';
+
+class TryGuessingWebFingerInteractor {
+  const TryGuessingWebFingerInteractor(this._authenticationOIDCRepository);
+
+  final AuthenticationOIDCRepository _authenticationOIDCRepository;
+
+  Stream<Either<Failure, Success>> execute(
```
Comment (hoangdat) [inline lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart:16]: should add the tests for it

File: lib/features/login/domain/model/oidc_guessing_origin.dart
Comment (tddang-linagora) [reply lib/features/login/domain/model/oidc_guessing_origin.dart:1]: Done

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart (hunk 9 above)
Comment (tddang-linagora) [reply lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart]: Updated

File: lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart (hunk 10 above)
Comment (tddang-linagora) [reply lib/features/login/domain/usecases/try_guessing_web_finger_interactor.dart:16]: Done

File: test/features/login/domain/usecases/try_guessing_web_finger_interactor_test.dart [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    verifyInOrder([
+      mockRepository.checkOIDCIsAvailable(oidcRequest1),
+      mockRepository.checkOIDCIsAvailable(oidcRequest2),
+    ]);
+    
+    // Verify second request was made but not completed
+    expect(completer2.isCompleted, isFalse);
+  });
+
+  test('should return failure when all responses are null', () async {
```
Comment (hoangdat) [inline test/features/login/domain/usecases/try_guessing_web_finger_interactor_test.dart:69]: add case `test('should still return success even one request failure', () async {`

**Review by hoangdat — APPROVED** [review]
